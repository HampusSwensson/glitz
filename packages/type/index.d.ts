import * as CSS from 'csstype';

export interface Style extends FeaturedProperties, PseudoMap {}

export interface Properties extends CSS.StandardLonghandPropertiesFallback, CSS.VendorPropertiesFallback {}

export interface FeaturedProperties
  extends Omit<Properties, keyof ExtendedProperties>,
    ExtendedProperties,
    Omit<CSS.StandardShorthandPropertiesFallback, keyof ShorthandProperties>,
    ShorthandProperties {
  '@keyframes'?: PropertiesList;
  '@font-face'?: FontFace;
}

export interface ExtendedProperties {
  animationName?: PropertiesList | Properties['animationName'];
  fontFamily?: FontFamily;
}

export interface ShorthandProperties {
  animation?: Animation;
  background?: Background;
  border?: Border;
  borderBottom?: BorderBottom;
  borderImage?: BorderImage;
  borderLeft?: BorderLeft;
  borderRight?: BorderRight;
  borderTop?: BorderTop;
  flex?: Flex;
  font?: Font;
  grid?: Grid;
  maskBorder?: MaskBorder;
  mask?: Mask;
  margin?: Margin;
  offset?: Offset;
  outline?: Outline;
  padding?: Padding;
  transition?: Transition;
}

interface Animation {
  delay?: CSS.StandardLonghandPropertiesFallback['animationDelay'];
  direction?: CSS.StandardLonghandPropertiesFallback['animationDirection'];
  duration?: CSS.StandardLonghandPropertiesFallback['animationDuration'];
  fillMode?: CSS.StandardLonghandPropertiesFallback['animationFillMode'];
  iterationCount?: CSS.StandardLonghandPropertiesFallback['animationIterationCount'];
  name?: PropertiesList | CSS.StandardLonghandPropertiesFallback['animationName'];
  playState?: CSS.StandardLonghandPropertiesFallback['animationPlayState'];
  timingFunction?: CSS.StandardLonghandPropertiesFallback['animationTimingFunction'];
}

interface Background {
  attachment?: CSS.StandardLonghandPropertiesFallback['backgroundAttachment'];
  blendMode?: CSS.StandardLonghandPropertiesFallback['backgroundBlendMode'];
  clip?: CSS.StandardLonghandPropertiesFallback['backgroundClip'];
  color?: CSS.StandardLonghandPropertiesFallback['backgroundColor'];
  image?: CSS.StandardLonghandPropertiesFallback['backgroundImage'];
  origin?: CSS.StandardLonghandPropertiesFallback['backgroundOrigin'];
  position?: CSS.StandardLonghandPropertiesFallback['backgroundPosition'];
  repeat?: CSS.StandardLonghandPropertiesFallback['backgroundRepeat'];
  size?: CSS.StandardLonghandPropertiesFallback['backgroundSize'];
}

interface Border {
  collapse?: CSS.StandardLonghandPropertiesFallback['borderCollapse'];
  color?: CSS.StandardShorthandPropertiesFallback['borderColor'];
  blockEnd?: CSS.StandardShorthandPropertiesFallback['borderBlockEnd'];
  blockStart?: CSS.StandardShorthandPropertiesFallback['borderBlockStart'];
  inlineEnd?: CSS.StandardShorthandPropertiesFallback['borderInlineEnd'];
  inlineStart?: CSS.StandardShorthandPropertiesFallback['borderInlineStart'];
  radius?: CSS.StandardShorthandPropertiesFallback['borderRadius'];
  spacing?: CSS.StandardLonghandPropertiesFallback['borderSpacing'];
  style?: CSS.StandardShorthandPropertiesFallback['borderStyle'];
  width?: CSS.StandardShorthandPropertiesFallback['borderWidth'];
}

interface BorderBottom {
  color?: CSS.StandardLonghandPropertiesFallback['borderBottomColor'];
  style?: CSS.StandardLonghandPropertiesFallback['borderBottomStyle'];
  width?: CSS.StandardLonghandPropertiesFallback['borderBottomWidth'];
}

interface BorderImage {
  outset?: CSS.StandardLonghandPropertiesFallback['borderImageOutset'];
  repeat?: CSS.StandardLonghandPropertiesFallback['borderImageRepeat'];
  slice?: CSS.StandardLonghandPropertiesFallback['borderImageSlice'];
  source?: CSS.StandardLonghandPropertiesFallback['borderImageSource'];
  width?: CSS.StandardLonghandPropertiesFallback['borderImageWidth'];
}

interface BorderLeft {
  color?: CSS.StandardLonghandPropertiesFallback['borderLeftColor'];
  style?: CSS.StandardLonghandPropertiesFallback['borderLeftStyle'];
  width?: CSS.StandardLonghandPropertiesFallback['borderLeftWidth'];
}

interface BorderRight {
  color?: CSS.StandardLonghandPropertiesFallback['borderRightColor'];
  style?: CSS.StandardLonghandPropertiesFallback['borderRightStyle'];
  width?: CSS.StandardLonghandPropertiesFallback['borderRightWidth'];
}

interface BorderTop {
  color?: CSS.StandardLonghandPropertiesFallback['borderTopColor'];
  style?: CSS.StandardLonghandPropertiesFallback['borderTopStyle'];
  width?: CSS.StandardLonghandPropertiesFallback['borderTopWidth'];
}

interface Flex {
  basis?: CSS.StandardLonghandPropertiesFallback['flexBasis'];
  direction?: CSS.StandardLonghandPropertiesFallback['flexDirection'];
  grow?: CSS.StandardLonghandPropertiesFallback['flexGrow'];
  shrink?: CSS.StandardLonghandPropertiesFallback['flexShrink'];
  wrap?: CSS.StandardLonghandPropertiesFallback['flexWrap'];
}

interface Font {
  family?: FontFamily;
  featureSettings?: CSS.StandardLonghandPropertiesFallback['fontFeatureSettings'];
  kerning?: CSS.StandardLonghandPropertiesFallback['fontKerning'];
  languageOverride?: CSS.StandardLonghandPropertiesFallback['fontLanguageOverride'];
  variationSettings?: CSS.StandardLonghandPropertiesFallback['fontVariationSettings'];
  size?: CSS.StandardLonghandPropertiesFallback['fontSize'];
  sizeAdjust?: CSS.StandardLonghandPropertiesFallback['fontSizeAdjust'];
  stretch?: CSS.StandardLonghandPropertiesFallback['fontStretch'];
  style?: CSS.StandardLonghandPropertiesFallback['fontStyle'];
  synthesis?: CSS.StandardLonghandPropertiesFallback['fontSynthesis'];
  variant?: CSS.StandardLonghandPropertiesFallback['fontVariant'];
  weight?: CSS.StandardLonghandPropertiesFallback['fontWeight'];
}

interface Grid {
  autoColumns?: CSS.StandardLonghandPropertiesFallback['gridAutoColumns'];
  autoFlow?: CSS.StandardLonghandPropertiesFallback['gridAutoFlow'];
  autoRows?: CSS.StandardLonghandPropertiesFallback['gridAutoRows'];
  column?: CSS.StandardShorthandPropertiesFallback['gridColumn'];
  columnGap?: CSS.StandardLonghandPropertiesFallback['gridColumnGap'];
  row?: CSS.StandardShorthandPropertiesFallback['gridRow'];
  rowGap?: CSS.StandardLonghandPropertiesFallback['gridRowGap'];
  template?: CSS.StandardShorthandPropertiesFallback['gridTemplate'];
}

interface MaskBorder {
  mode?: CSS.StandardLonghandPropertiesFallback['maskBorderMode'];
  outset?: CSS.StandardLonghandPropertiesFallback['maskBorderOutset'];
  repeat?: CSS.StandardLonghandPropertiesFallback['maskBorderRepeat'];
  slice?: CSS.StandardLonghandPropertiesFallback['maskBorderSlice'];
  source?: CSS.StandardLonghandPropertiesFallback['maskBorderSource'];
  width?: CSS.StandardLonghandPropertiesFallback['maskBorderWidth'];
}

interface Mask {
  clip?: CSS.StandardLonghandPropertiesFallback['maskClip'];
  composite?: CSS.StandardLonghandPropertiesFallback['maskComposite'];
  image?: CSS.StandardLonghandPropertiesFallback['maskImage'];
  mode?: CSS.StandardLonghandPropertiesFallback['maskMode'];
  origin?: CSS.StandardLonghandPropertiesFallback['maskOrigin'];
  position?: CSS.StandardLonghandPropertiesFallback['maskPosition'];
  repeat?: CSS.StandardLonghandPropertiesFallback['maskRepeat'];
  size?: CSS.StandardLonghandPropertiesFallback['maskSize'];
  type?: CSS.StandardLonghandPropertiesFallback['maskType'];
}

interface Margin {
  x?: CSS.StandardLonghandPropertiesFallback['marginLeft'];
  y?: CSS.StandardLonghandPropertiesFallback['marginTop'];
  top?: CSS.StandardLonghandPropertiesFallback['marginTop'];
  right?: CSS.StandardLonghandPropertiesFallback['marginRight'];
  bottom?: CSS.StandardLonghandPropertiesFallback['marginBottom'];
  left?: CSS.StandardLonghandPropertiesFallback['marginLeft'];
}

interface Offset {
  anchor?: CSS.StandardLonghandPropertiesFallback['offsetAnchor'];
  blockEnd?: CSS.StandardLonghandPropertiesFallback['offsetBlockEnd'];
  blockStart?: CSS.StandardLonghandPropertiesFallback['offsetBlockStart'];
  inlineEnd?: CSS.StandardLonghandPropertiesFallback['offsetInlineEnd'];
  inlineStart?: CSS.StandardLonghandPropertiesFallback['offsetInlineStart'];
  distance?: CSS.StandardLonghandPropertiesFallback['offsetDistance'];
  path?: CSS.StandardLonghandPropertiesFallback['offsetPath'];
  position?: CSS.StandardLonghandPropertiesFallback['offsetPosition'];
  rotate?: CSS.StandardLonghandPropertiesFallback['offsetRotate'];
}

interface Outline {
  color?: CSS.StandardLonghandPropertiesFallback['outlineColor'];
  offset?: CSS.StandardLonghandPropertiesFallback['outlineOffset'];
  style?: CSS.StandardLonghandPropertiesFallback['outlineStyle'];
  width?: CSS.StandardLonghandPropertiesFallback['outlineWidth'];
}

interface Padding {
  x?: CSS.StandardLonghandPropertiesFallback['paddingLeft'];
  y?: CSS.StandardLonghandPropertiesFallback['paddingTop'];
  top?: CSS.StandardLonghandPropertiesFallback['paddingTop'];
  right?: CSS.StandardLonghandPropertiesFallback['paddingRight'];
  bottom?: CSS.StandardLonghandPropertiesFallback['paddingBottom'];
  left?: CSS.StandardLonghandPropertiesFallback['paddingLeft'];
}

interface Transition {
  delay?: CSS.StandardLonghandPropertiesFallback['transitionDelay'];
  duration?: CSS.StandardLonghandPropertiesFallback['transitionDuration'];
  property?: CSS.StandardLonghandPropertiesFallback['transitionProperty'];
  timingFunction?: CSS.StandardLonghandPropertiesFallback['transitionTimingFunction'];
}

export type PseudoMap = { [P in CSS.SimplePseudos]?: FeaturedProperties & PseudoMap };

export interface PropertiesList {
  [identifier: string]: Properties;
}

export type FontFamily =
  | FontFace
  | CSS.StandardLonghandProperties['fontFamily']
  | Array<FontFace | CSS.StandardLonghandProperties['fontFamily']>;

export type FontFace = Omit<CSS.FontFaceFallback, 'fontFamily'>;

type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];
type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;
