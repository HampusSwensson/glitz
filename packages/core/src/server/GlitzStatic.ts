import { Style } from '@glitz/type';
import Base from '../core/Base';
import { Options } from '../types/options';
import { createHashCounter } from '../utils/hash';
import InjectorServer from './InjectorServer';

export default class GlitzStatic<TStyle = Style> extends Base<TStyle> {
  public getStyle: () => string;
  constructor(options: Options = {}) {
    const prefix = options.prefix;
    const classHasher = createHashCounter(prefix);
    const keyframesHasher = createHashCounter(prefix);

    let plain: InjectorServer;
    const mediaIndex: {
      [media: string]: InjectorServer;
    } = {};

    const injector = (media?: string) =>
      media
        ? (mediaIndex[media] = mediaIndex[media] || new InjectorServer(classHasher, keyframesHasher))
        : (plain = plain || new InjectorServer(classHasher, keyframesHasher));

    super(injector, options.transformer, options.atomic);

    this.getStyle = () => {
      let css = '';
      if (plain) {
        css += plain.getStyle();
      }
      const medias = options.mediaOrder ? Object.keys(mediaIndex).sort(options.mediaOrder) : Object.keys(mediaIndex);
      for (const media of medias) {
        css += `@media ${media}{${mediaIndex[media].getStyle()}}`;
      }
      return css;
    };
  }
}
