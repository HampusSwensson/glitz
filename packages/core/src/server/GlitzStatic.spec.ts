import { ResolvedDeclarations, CommonValue, Style } from '@glitz/type';
import GlitzStatic from './GlitzStatic';

interface TestStyle extends Style {
  '@media (min-width: 100px)'?: Style;
  '@media (min-width: 200px)'?: Style;
  '@media (min-width: 300px)'?: Style;
  '@media (min-width: 768px)'?: Style;
  '@media (min-width: 992px)'?: Style;
}

describe('server', () => {
  it('injects plain rule', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(server.injectStyle({ color: 'red' })).toBe('a');
    expect(server.getStyle()).toMatchInlineSnapshot(`".a{color:red}"`);
  });
  it('injects shorthand rule', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(
      server.injectStyle({
        padding: { left: '10px', right: '10px', top: '10px', bottom: '10px' },
      }),
    ).toBe('a b c d');

    expect(
      server.injectStyle({
        grid: { column: { gap: '10px' } },
      }),
    ).toBe('e');

    expect(
      server.injectStyle({
        margin: { x: '10px' },
      }),
    ).toBe('f g');

    expect(
      server.injectStyle({
        margin: { y: '10px' },
      }),
    ).toBe('h i');

    expect(
      server.injectStyle({
        margin: { xy: '20px' },
      }),
    ).toBe('j k l m');

    expect(
      server.injectStyle({
        padding: { left: '20px' },
        paddingLeft: '30px',
      }),
    ).toBe('n');

    expect(
      server.injectStyle({
        animationName: { from: { padding: { left: '20px' } } },
      }),
    ).toBe('o');

    expect(
      server.injectStyle({
        fontFamily: { fontFamily: 'x' },
      }),
    ).toBe('p');

    expect(
      server.injectStyle({
        border: { xy: { width: 0 }, x: { color: 'red' }, y: { color: 'green' } },
      }),
    ).toBe('q r s t u v w x');

    expect(
      server.injectStyle({
        border: { y: { x: { radius: 0 } } },
      }),
    ).toBe('y z a0 a1');

    expect(
      server.injectStyle({
        border: { x: { width: '10px' }, left: { width: '20px' } },
      }),
    ).toBe('a2 a3');

    expect(server.getStyle()).toMatchInlineSnapshot(
      `"@font-face{font-family:x}@keyframes a{from{padding-left:20px}}.a{padding-bottom:10px}.b{padding-top:10px}.c{padding-right:10px}.d{padding-left:10px}.e{grid-column-gap:10px}.f{margin-right:10px}.g{margin-left:10px}.h{margin-bottom:10px}.i{margin-top:10px}.j{margin-bottom:20px}.k{margin-top:20px}.l{margin-right:20px}.m{margin-left:20px}.n{padding-left:30px}.o{animation-name:a}.p{font-family:x}.q{border-bottom-color:green}.r{border-top-color:green}.s{border-right-color:red}.t{border-left-color:red}.u{border-bottom-width:0}.v{border-top-width:0}.w{border-right-width:0}.x{border-left-width:0}.y{border-bottom-right-radius:0}.z{border-bottom-left-radius:0}.a0{border-top-right-radius:0}.a1{border-top-left-radius:0}.a2{border-left-width:20px}.a3{border-right-width:10px}"`,
    );
  });
  it('injects pseudo selector', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(server.injectStyle({ color: 'red' })).toBe('a');
    expect(server.injectStyle({ ':hover': { color: 'red' } })).toBe('b');
    expect(server.getStyle()).toMatchInlineSnapshot(`".a{color:red}.b:hover{color:red}"`);
  });
  it('injects nested pseudo selector', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(server.injectStyle({ ':first-child': { ':hover': { color: 'red' } } })).toBe('a');
    expect(server.getStyle()).toMatchInlineSnapshot(`".a:first-child:hover{color:red}"`);
  });
  it('injects attribute selector', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(server.injectStyle({ color: 'red' })).toBe('a');
    expect(server.injectStyle({ '[disabled]': { color: 'red' } })).toBe('b');
    expect(server.getStyle()).toMatchInlineSnapshot(`".a{color:red}.b[disabled]{color:red}"`);
  });
  it('injects nested attribute selector', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(server.injectStyle({ '[readonly]': { '[disabled]': { color: 'red' } } })).toBe('a');
    expect(server.getStyle()).toMatchInlineSnapshot(`".a[readonly][disabled]{color:red}"`);
  });
  it('injects media rule', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(server.injectStyle({ color: 'red' })).toBe('a');
    expect(server.injectStyle({ ':hover': { color: 'red' } })).toBe('b');
    expect(server.injectStyle({ '@media (min-width: 768px)': { color: 'red' } })).toBe('c');
    expect(server.injectStyle({ '@media (min-width: 768px)': { ':hover': { color: 'red' } } })).toBe('d');
    expect(server.getStyle()).toMatchInlineSnapshot(
      `".a{color:red}.b:hover{color:red}@media (min-width: 768px){.c{color:red}.d:hover{color:red}}"`,
    );
  });
  it('injects media markup in certain order', () => {
    const server = new GlitzStatic<TestStyle>({
      mediaOrder: (a: string, b: string) => {
        if (a < b) {
          return -1;
        }
        if (a > b) {
          return 1;
        }
        return 0;
      },
    });

    expect(server.injectStyle({ '@media (min-width: 300px)': { color: 'red' } })).toBe('a');
    expect(server.injectStyle({ '@media (min-width: 100px)': { color: 'red' } })).toBe('b');
    expect(server.injectStyle({ '@media (min-width: 200px)': { color: 'red' } })).toBe('c');
    expect(server.injectStyle({ color: 'red' })).toBe('d');

    expect(server.getStyle()).toMatchInlineSnapshot(
      `".d{color:red}@media (min-width: 100px){.b{color:red}}@media (min-width: 200px){.c{color:red}}@media (min-width: 300px){.a{color:red}}"`,
    );
  });
  it('injects atomic rules', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(
      server.injectStyle({
        color: 'red',
        background: { color: 'green' },
        borderLeftColor: 'blue',
        ':hover': { color: 'red', background: { color: 'green' }, borderLeftColor: 'blue' },
        '@media (min-width: 768px)': { color: 'red', background: { color: 'green' }, borderLeftColor: 'blue' },
      }),
    ).toBe('a b c d e f g h i');

    expect(server.getStyle()).toMatchInlineSnapshot(
      `".g{border-left-color:blue}.h{background-color:green}.i{color:red}.d:hover{border-left-color:blue}.e:hover{background-color:green}.f:hover{color:red}@media (min-width: 768px){.a{border-left-color:blue}.b{background-color:green}.c{color:red}}"`,
    );
  });
  it('injects keyframes rule', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(server.injectStyle({ '@keyframes': { from: { color: 'red' }, to: { color: 'green' } } })).toBe('a');
    expect(server.injectStyle({ animationName: { from: { color: 'blue' }, to: { color: 'white' } } })).toBe('b');
    expect(server.injectStyle({ animation: { name: { from: { color: 'blue' }, to: { color: 'white' } } } })).toBe('b');
    expect(server.injectStyle({ animationName: 'some-thing' })).toBe('c');
    expect(server.getStyle()).toMatchInlineSnapshot(
      `"@keyframes a{from{color:red}to{color:green}}@keyframes b{from{color:blue}to{color:white}}.a{animation-name:a}.b{animation-name:b}.c{animation-name:some-thing}"`,
    );
  });
  it('injects font face rule', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(
      server.injectStyle({
        '@font-face': {
          fontFamily: 'x',
          fontStyle: 'normal',
          fontWeight: 400,
          src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2')",
        },
      }),
    ).toBe('a');

    expect(
      server.injectStyle({
        fontFamily: {
          fontFamily: 'y',
          fontStyle: 'normal',
          fontWeight: 400,
          src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
        },
      }),
    ).toBe('b');

    expect(
      server.injectStyle({
        font: {
          family: {
            fontFamily: 'y',
            fontStyle: 'normal',
            fontWeight: 400,
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
          },
        },
      }),
    ).toBe('b');

    expect(
      server.injectStyle({
        fontFamily: [
          {
            fontFamily: 'x',
            fontStyle: 'normal',
            fontWeight: 400,
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
          },
          'sans-serif',
        ],
      }),
    ).toBe('c');

    expect(
      server.injectStyle({
        font: {
          family: [
            {
              fontFamily: 'x',
              fontStyle: 'normal',
              fontWeight: 400,
              src:
                "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
            },
            'sans-serif',
          ],
        },
      }),
    ).toBe('c');

    expect(
      server.injectStyle({
        fontFamily: 'sans-serif',
      }),
    ).toBe('d');

    expect(server.getStyle()).toMatchInlineSnapshot(
      `"@font-face{font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2');font-family:x}@font-face{font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2');font-family:y}@font-face{font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2');font-family:x}.a{font-family:x}.b{font-family:y}.c{font-family:x,sans-serif}.d{font-family:sans-serif}"`,
    );
  });
  it('injects different combinations', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(
      server.injectStyle({
        color: 'red',
        '@media (min-width: 768px)': { color: 'green' },
        '@media (min-width: 992px)': { color: 'blue' },
      }),
    ).toBe('a b c');
    expect(server.getStyle()).toMatchInlineSnapshot(
      `".c{color:red}@media (min-width: 992px){.a{color:blue}}@media (min-width: 768px){.b{color:green}}"`,
    );
  });
  it('injects rule deeply', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(server.injectStyle([{ color: 'green' }, { color: 'red' }])).toBe('a');

    expect(
      server.injectStyle([
        {
          padding: { left: '10px' },
        },
        {
          paddingLeft: '20px',
        },
      ]),
    ).toBe('b');

    expect(server.injectStyle([{ ':hover': { color: 'green' } }, { ':hover': { color: 'red' } }])).toBe('c');
    expect(
      server.injectStyle([
        { ':first-child': { ':hover': { color: 'green' } } },
        { ':first-child': { ':hover': { color: 'red' } } },
      ]),
    ).toBe('d');

    expect(
      server.injectStyle([
        { '@keyframes': { from: { color: 'red' }, to: { color: 'green' } } },
        { '@keyframes': { from: { color: 'green' }, to: { color: 'blue' } } },
      ]),
    ).toBe('e');
    expect(
      server.injectStyle([
        { animationName: { from: { color: 'blue' }, to: { color: 'white' } } },
        { animationName: { from: { color: 'white' }, to: { color: 'black' } } },
      ]),
    ).toBe('f');

    expect(
      server.injectStyle([
        {
          '@font-face': {
            fontFamily: 'x',
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2')",
          },
        },
        {
          '@font-face': {
            fontFamily: 'y',
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
          },
        },
      ]),
    ).toBe('g');
    expect(
      server.injectStyle([
        {
          '@font-face': {
            fontFamily: 'y',
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
          },
        },
        {
          '@font-face': {
            fontFamily: 'z',
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2')",
          },
        },
      ]),
    ).toBe('h');

    expect(
      server.injectStyle([
        { '@media (min-width: 768px)': { color: 'green' } },
        { '@media (min-width: 768px)': { color: 'red' } },
      ]),
    ).toBe('i');
    expect(
      server.injectStyle([
        { '@media (min-width: 768px)': { ':hover': { color: 'green' } } },
        { '@media (min-width: 768px)': { ':hover': { color: 'red' } } },
      ]),
    ).toBe('j');

    expect(server.getStyle()).toMatchInlineSnapshot(
      `"@font-face{src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2');font-family:y}@font-face{src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2');font-family:z}@keyframes a{from{color:green}to{color:blue}}@keyframes b{from{color:white}to{color:black}}.a{color:red}.b{padding-left:20px}.e{animation-name:a}.f{animation-name:b}.g{font-family:y}.h{font-family:z}.c:hover{color:red}.d:first-child:hover{color:red}@media (min-width: 768px){.i{color:red}.j:hover{color:red}}"`,
    );
  });
  it('deletes properties', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(
      server.injectStyle({
        color: 'red',
        paddingRight: '10px',
        padding: { right: undefined },
        animationName: { from: { color: 'red' }, to: { color: 'green' } },
        animation: { name: undefined },
      }),
    ).toBe('a');

    expect(
      server.injectStyle([
        {
          color: 'red',
          paddingRight: '10px',
        },
        {
          paddingRight: (null as any) as undefined,
        },
      ]),
    ).toBe('a');

    expect(server.getStyle()).toMatchInlineSnapshot(`".a{color:red}"`);
  });
  it('applies transformer', () => {
    const server = new GlitzStatic<TestStyle>({
      transformer: properties => {
        const prefixed: ResolvedDeclarations = {};
        for (const property in properties) {
          const value: CommonValue = properties[property];
          if (property === 'appearance' && value === 'none') {
            prefixed.MozAppearance = value;
          }
          prefixed[property] = properties[property];
        }
        return prefixed;
      },
    });

    expect(server.injectStyle({ appearance: 'none', animationName: { from: { appearance: 'none' } } })).toBe('a b');
    expect(server.getStyle()).toMatchInlineSnapshot(
      `"@keyframes a{from{-moz-appearance:none;appearance:none}}.a{animation-name:a}.b{-moz-appearance:none;appearance:none}"`,
    );
  });
  it('passes theme', () => {
    const server = new GlitzStatic<TestStyle>();

    expect(server.injectStyle({ color: (theme: any) => theme.text }, { text: 'red' })).toBe('a');
    expect(server.getStyle()).toMatchInlineSnapshot(`".a{color:red}"`);
  });
});
