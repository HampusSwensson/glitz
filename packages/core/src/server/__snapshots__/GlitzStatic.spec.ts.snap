// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server applies transformer 1`] = `"@keyframes a{from{-moz-appearance:none;appearance:none}}.a{-moz-appearance:none;appearance:none}.b{animation-name:a}"`;

exports[`server deletes properties 1`] = `".a{color:red}"`;

exports[`server injects atomic rules 1`] = `".a{color:red}.b{background-color:green}.c{border-left-color:blue}.d:hover{color:red}.e:hover{background-color:green}.f:hover{border-left-color:blue}@media (min-width: 768px){.g{color:red}.h{background-color:green}.i{border-left-color:blue}}"`;

exports[`server injects attribute selector 1`] = `".a{color:red}.b[disabled]{color:red}"`;

exports[`server injects different combinations 1`] = `".a{color:red}@media (min-width: 768px){.b{color:green}}@media (min-width: 992px){.c{color:blue}}"`;

exports[`server injects font face rule 1`] = `"@font-face{font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2');font-family:x}@font-face{font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2');font-family:y}@font-face{font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2');font-family:x}.a{font-family:x}.b{font-family:y}.c{font-family:x,sans-serif}.d{font-family:sans-serif}"`;

exports[`server injects keyframes rule 1`] = `"@keyframes a{from{color:red}to{color:green}}@keyframes b{from{color:blue}to{color:white}}.a{animation-name:a}.b{animation-name:b}.c{animation-name:some-thing}"`;

exports[`server injects media markup in certain order 1`] = `".d{color:red}@media (min-width: 100px){.b{color:red}}@media (min-width: 200px){.c{color:red}}@media (min-width: 300px){.a{color:red}}"`;

exports[`server injects media rule 1`] = `".a{color:red}.b:hover{color:red}@media (min-width: 768px){.c{color:red}.d:hover{color:red}}"`;

exports[`server injects nested attribute selector 1`] = `".a[readonly][disabled]{color:red}"`;

exports[`server injects nested pseudo selector 1`] = `".a:first-child:hover{color:red}"`;

exports[`server injects non-atomic rules 1`] = `".a{color:red;background-color:green;border-left-color:blue}.b:hover{color:red;background-color:green;border-left-color:blue}@media (min-width: 768px){.c{color:red;background-color:green;border-left-color:blue}}"`;

exports[`server injects plain rule 1`] = `".a{color:red}"`;

exports[`server injects pseudo selector 1`] = `".a{color:red}.b:hover{color:red}"`;

exports[`server injects rule deeply 1`] = `"@font-face{src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2');font-family:y}@font-face{src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2');font-family:z}@keyframes a{from{color:green}to{color:blue}}@keyframes b{from{color:white}to{color:black}}.a{color:red}.b{padding-left:20px}.e{animation-name:a}.f{animation-name:b}.g{font-family:y}.h{font-family:z}.c:hover{color:red}.d:first-child:hover{color:red}@media (min-width: 768px){.i{color:red}.j:hover{color:red}}"`;

exports[`server injects shorthand rule 1`] = `"@font-face{font-family:x}@keyframes a{from{padding-left:20px}}.a{padding-left:10px}.b{padding-right:10px}.c{padding-top:10px}.d{padding-bottom:10px}.e{grid-column-gap:10px}.f{margin-left:10px}.g{margin-right:10px}.h{margin-top:10px}.i{margin-bottom:10px}.j{margin-left:20px}.k{margin-right:20px}.l{margin-top:20px}.m{margin-bottom:20px}.n{padding-left:30px}.o{animation-name:a}.p{font-family:x}.q{border-left-width:0}.r{border-right-width:0}.s{border-top-width:0}.t{border-bottom-width:0}.u{border-left-color:red}.v{border-right-color:red}.w{border-top-color:green}.x{border-bottom-color:green}.y{border-top-left-radius:0}.z{border-top-right-radius:0}.a0{border-bottom-left-radius:0}.a1{border-bottom-right-radius:0}.a2{border-right-width:10px}.a3{border-left-width:20px}"`;

exports[`server passes theme 1`] = `".a{color:red}"`;

exports[`server preserves order 1`] = `".a{padding-top:20px}.b{color:red}.c{padding-right:30px}.e{padding-left:30px}.d:hover{color:green}"`;

exports[`server preserves order 2`] = `".a{padding-top:20px;color:red;padding-right:30px;padding-left:30px}.b:hover{color:green}"`;
